src/runner.rs

```diff
@@ -70,14 +87,66 @@ impl<'a, E: Externals> Interpreter<'a, E> {
 				)?;
 		}
 
-		let initial_frame = FunctionContext::new(func.clone());
-
 		let mut call_stack = Vec::new();
+		let initial_frame = FunctionContext::new(func.clone());
 		call_stack.push(initial_frame);
 
-		self.run_interpreter_loop(&mut call_stack)?;
+		let return_type = func.signature().return_type();
+
+		Ok(Interpreter {
+			value_stack,
+			call_stack,
+			return_type,
+			state: InterpreterState::Initialized,
+		})
+	}
+
+	pub fn state(&self) -> &InterpreterState {
+		&self.state
+	}
+
+	pub fn start_execution<'a, E: Externals + 'a>(&mut self, externals: &'a mut E) -> Result<Option<RuntimeValue>, Trap> {
+		// Ensure that the VM has not been executed.
+		assert!(self.state == InterpreterState::Initialized);
```

Comment:
Not sure about all those asserts

This all supposed to be called in places with zero tolerance to panics

Can we at least write why this assert will always pass?

Reply:
The public interface will not panic. In `func` mod's `FuncInvocation::start_execution`, we check whether the state is `Initialized` first before calling `Interpreter::start_execution`. But if there's an internal incorrect usage in our code, then panic do happen. Do we prefer to return additional errors here or is this panic okay?

Reply:
Oh, an explanation is ok